sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp

```diff
@@ -0,0 +1,192 @@
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/raw_ostream.h"
+#include <iostream>
+#include <sstream>
+
+using namespace jit_compiler;
+using namespace jit_compiler::translation;
+using namespace llvm;
+
+void SPIRVLLVMTranslator::getAttributeValues(std::vector<std::string> &Values,
+                                             MDNode *MD, size_t NumValues) {
+  assert(MD->getNumOperands() == NumValues && "Incorrect number of values");
+  for (const auto &MDOp : MD->operands()) {
+    auto *ConstantMD = dyn_cast<ConstantAsMetadata>(MDOp);
+    assert(ConstantMD && "Metadata not constant");
+    auto *ConstantVal = ConstantMD->getValue();
+    auto *ConstInt = dyn_cast<ConstantInt>(ConstantVal);
+    assert(ConstInt && "Not a constant integer");
+    Values.push_back(std::to_string(ConstInt->getZExtValue()));
+  }
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+static const char *REQD_WORK_GROUP_SIZE_ATTR = "reqd_work_group_size";
+// NOLINTNEXTLINE(readability-identifier-naming)
+static const char *WORK_GROUP_SIZE_HINT_ATTR = "work_group_size_hint";
+
+void SPIRVLLVMTranslator::restoreKernelAttributes(Module *Mod,
+                                                  SYCLKernelInfo &Info) {
+  auto *KernelFunction = Mod->getFunction(Info.Name);
+  assert(KernelFunction && "Kernel function not present in module");
+  if (KernelFunction->hasMetadata(REQD_WORK_GROUP_SIZE_ATTR)) {
+    auto *MD = KernelFunction->getMetadata(REQD_WORK_GROUP_SIZE_ATTR);
+    SYCLKernelAttribute ReqdAttr{REQD_WORK_GROUP_SIZE_ATTR};
+    getAttributeValues(ReqdAttr.Values, MD, 3);
+    Info.Attributes.push_back(ReqdAttr);
+  }
+  if (KernelFunction->hasMetadata(WORK_GROUP_SIZE_HINT_ATTR)) {
+    auto *MD = KernelFunction->getMetadata(WORK_GROUP_SIZE_HINT_ATTR);
+    SYCLKernelAttribute HintAttr{WORK_GROUP_SIZE_HINT_ATTR};
+    getAttributeValues(HintAttr.Values, MD, 3);
+    Info.Attributes.push_back(HintAttr);
+  }
+}
+
+SPIRV::TranslatorOpts &SPIRVLLVMTranslator::translatorOpts() {
+  static auto Opts = []() -> SPIRV::TranslatorOpts {
+    // Options for translation between SPIR-V and LLVM IR.
+    // Set SPIRV-V 1.2 as the maximum version number for now.
+    SPIRV::TranslatorOpts TransOpt{SPIRV::VersionNumber::SPIRV_1_2};
+    // Enable attachment of kernel arg names as metadata.
+    TransOpt.enableGenArgNameMD();
+    // Enable mem2reg.
+    TransOpt.setMemToRegEnabled(true);
+    // Enable all extensions.
+    // TODO: Specifically enable only the
+    // extensions listed in the KernelInfo.
+    // FIXME: Because there's no size provided,
+    // there's currently no obvious way to iterate the
+    // array of extensions in KernelInfo.
+    TransOpt.enableAllExtensions();
+    TransOpt.setDesiredBIsRepresentation(
+        SPIRV::BIsRepresentation::SPIRVFriendlyIR);
+    // TODO: Do we need to care about specialization constants?
```

Comment:
If they were present in the original program, then yes, we must care about them. SYCL code may set certain values to them and during SPIR-V -> LLVM IR translation spec constants will be replaced with a regular constants. Default initialization will be used if no values provided. This replacement is a non-reversible operation, we won't preserve any information to make those spec constants re-configurable later

Reply:
Yes, we definitely need to handle them. At the point where the SYCL runtime calls the JIT, the values for specialization constants should be known, so we could replace them with the user-specified value. Alternatively, we would need to make sure the spec constants are re-configurable in the output of the JIT compilation process and have the driver perform the instantiation. 

I updated the TODO to reflect this, so we can address it in a later PR.